File: Trees_Graphs/graph.rb
class Graph
  def initialize(nodes = [])
    @nodes = nodes
  end

  def createNode(value, connections)
    @nodes << Node.new(value, connections)
  end

  def addNode(node)
    @nodes << node
  end
end

class Node
  def initialize(value, connections = [])
    @value = value
    @connections = connections
  end

  def add_connection(node)
    @connections << node
  end
end


a = Node.new('a')
b = Node.new('b')
c = Node.new('c')
d = Node.new('d')
e = Node.new('e')
f = Node.new('f')

a.add_connection(b)
a.add_connection(d)
b.add_connection(e)
d.add_connection(c)
e.add_connection(f)

directed_non_Circular = Graph.new([a,b,c,d,e,f])


g = Node.new('g')
h = Node.new('h')
i = Node.new('i')
j = Node.new('j')
k = Node.new('k')

g.add_connection(h)
g.add_connection(j)
h.add_connection(i)
i.add_connection(g)
j.add_connection(k)
k.add_connection(i)

directed_circular = Graph.new([g,h,i,j,k])

l = Node.new('l')
m = Node.new('m')
n = Node.new('n')
o = Node.new('o')
p = Node.new('p')
q = Node.new('q')
r = Node.new('r')

l.add_connection(m)
m.add_connection(l)
m.add_connection(r)
m.add_connection(n)
r.add_connection(m)
n.add_connection(m)
n.add_connection(q)
q.add_connection(n)
n.add_connection(o)
o.add_connection(n)
o.add_connection(p)
p.add_connection(o)


undirected_non_circular = Graph.new([l,m,n,o,p,q,r])

s = Node.new('s')
t = Node.new('t')
u = Node.new('u')
v = Node.new('v')
w = Node.new('w')
x = Node.new('x')
y = Node.new('y')
z = Node.new('z')

x.add_connection(s)
x.add_connection(y)
s.add_connection(x)
y.add_connection(x)
y.add_connection(v)
y.add_connection(s)
s.add_connection(y)
v.add_connection(y)
v.add_connection(s)
v.add_connection(w)
s.add_connection(v)
w.add_connection(v)
w.add_connection(u)
u.add_connection(w)
u.add_connection(s)
s.add_connection(u)
u.add_connection(z)
z.add_connection(u)
u.add_connection(t)
t.add_connection(u)
z.add_connection(t)
t.add_connection(z)
t.add_connection(s)
s.add_connection(t)

undirected_circular = Graph.new([s,t,u,v,w,x,y,z])

File: Trees_Graphs/graph.js
class Graph {
  constructor(nodes = []) {
    this.nodes = nodes
  }

  createNode(value, connections) {
    const newNode = new Node(value, connections)
    this.nodes.push(newNode)
  }

  addNode(node) {
    this.nodes.push(node)
  }

}

class Node {
  constructor(value, connections = []) {
    this.value = value
    this.connections = connections
  }

  addConnection(node) {
    this.connections.push(node)
  }
}

class BSTNode {
  constructor(value, left = null, right = null) {
    this.value = value
    this.left = left
    this.right = right
  }

  printTree() {
    const printQueue = [this]

    while (printQueue.length > 0) {
      const currentNode = printQueue.shift()
      if (currentNode.left && currentNode.right) {
        printQueue.push(currentNode.left, currentNode.right)
      } else if (currentNode.left && !currentNode.right) {
        printQueue.push(currentNode.left)
      } else if(!currentNode.left && currentNode.right) {
        printQueue.push(currentNode.right)
      }
      console.log(currentNode.value)
    }
  }

}

class ModifedBSTNode extends BSTNode {
  constructor(value) {
    super(value)
    this.parent = null
  }
}

const a = new Node('a')
const b = new Node('b')
const c = new Node('c')
const d = new Node('d')
const e = new Node('e')
const f = new Node('f')

a.addConnection(b)
a.addConnection(d)
b.addConnection(e)
d.addConnection(c)
e.addConnection(f)

const directedNonCircular = new Graph([a,b,c,d,e,f])


const g = new Node('g')
const h = new Node('h')
const i = new Node('i')
const j = new Node('j')
const k = new Node('k')

g.addConnection(h)
g.addConnection(j)
h.addConnection(i)
i.addConnection(g)
j.addConnection(k)
k.addConnection(i)

const directedCircular = new Graph([g,h,i,j,k])

const l = new Node('l')
const m = new Node('m')
const n = new Node('n')
const o = new Node('o')
const p = new Node('p')
const q = new Node('q')
const r = new Node('r')

l.addConnection(m)
m.addConnection(l)
m.addConnection(r)
m.addConnection(n)
r.addConnection(m)
n.addConnection(m)
n.addConnection(q)
q.addConnection(n)
n.addConnection(o)
o.addConnection(n)
o.addConnection(p)
p.addConnection(o)


const undirectedNonCircular = new Graph([l,m,n,o,p,q,r])

const s = new Node('s')
const t = new Node('t')
const u = new Node('u')
const v = new Node('v')
const w = new Node('w')
const x = new Node('x')
const y = new Node('y')
const z = new Node('z')

x.addConnection(s)
x.addConnection(y)
s.addConnection(x)
y.addConnection(x)
y.addConnection(v)
y.addConnection(s)
s.addConnection(y)
v.addConnection(y)
v.addConnection(s)
v.addConnection(w)
s.addConnection(v)
w.addConnection(v)
w.addConnection(u)
u.addConnection(w)
u.addConnection(s)
s.addConnection(u)
u.addConnection(z)
z.addConnection(u)
u.addConnection(t)
t.addConnection(u)
z.addConnection(t)
t.addConnection(z)
t.addConnection(s)
s.addConnection(t)

const undirectedCircular = new Graph([s,t,u,v,w,x,y,z])

// Modified BST
const bstWithParent = new ModifedBSTNode(48)
const alpha = new ModifedBSTNode(35)
const bravo = new ModifedBSTNode(91)
const charlie = new ModifedBSTNode(21)
const delta = new ModifedBSTNode(55)
const echo = new ModifedBSTNode(92)
const foxtrot = new ModifedBSTNode(11)
const golf = new ModifedBSTNode(29)
const hotel = new ModifedBSTNode(77)
const inda = new ModifedBSTNode(71)

bstWithParent.left = alpha
bstWithParent.right = bravo
alpha.parent = bstWithParent
bravo.parent = bstWithParent

alpha.left = charlie
charlie.parent = alpha

charlie.left = foxtrot
charlie.right = golf
foxtrot.parent = charlie
golf.parent = charlie

bravo.left = delta
bravo.right = echo
delta.parent = bravo
echo.parent = bravo

delta.right = hotel
hotel.parent = delta

hotel.left = inda
inda.parent = hotel

module.exports = {
  directedNonCircular,
  directedCircular,
  undirectedNonCircular,
  undirectedCircular,
  BSTNode,
  ModifedBSTNode,
  bstWithParent
 }
